src/main/java/org/llorllale/cactoos/matchers/HasValues.java

```diff
@@ -53,43 +57,76 @@
  * @checkstyle ProtectedMethodInFinalClassCheck (200 lines)
  */
 public final class HasValues<X> extends TypeSafeDiagnosingMatcher<Iterable<X>> {
+
+    /**
+     * The expected values within the collection.
+     */
+    private final Collection<X> exp;
+
     /**
      * The function to check the {@link Iterable}.
      */
     private final UncheckedFunc<X, Boolean> fnc;
 
+    /**
+     * The function to describe the actual values in hamcrest terms.
+     */
+    private final UncheckedBiProc<Iterable<X>, Description> fact;
+
     /**
      * Ctor.
      * @param exp The expected values within unit test.
-     * @todo #32:30m Add the opportunity to print the exp values to the
-     *  description in order to avoid abstract message like
-     *  "The function applied to X,Y,Z is failed".
      */
     @SafeVarargs
     public HasValues(final X... exp) {
-        this(x -> new ListOf<>(exp).contains(x));
+        this(
+            new ListOf<>(exp),
+            x -> new ListOf<>(exp).contains(x),
+            (act, desc) -> desc.appendValue(new ListOf<>(act))
+        );
     }
 
     /**
      * Ctor.
      * @param fnc The function to check the {@link Iterable}.
      */
     public HasValues(final Func<X, Boolean> fnc) {
+        this(Collections.emptyList(), fnc,
+            (act, desc) -> desc.appendText("The function applied to ")
+                .appendValue(new ListOf<>(act))
+                .appendText(" is failed.")
+        );
+    }
+
+    /**
+     * Ctor.
+     * @param exp The expected values within the unit test.
+     * @param fnc The function to check the {@link Iterable}.
+     * @param fact The function to describe actual values in hamcrest terms.
```

Comment:
@dgroup "in hamcrest terms" sounds a bit unclear to me... could you elaborate ?

Reply:
@llorllale , agree, done.